app/assets/javascripts/discourse/app/lib/uppy-chunked-upload.js

```diff
@@ -0,0 +1,351 @@
+      const state = this.chunkState[i];
+      // eslint-disable-next-line no-continue
+      if (state.done || state.busy) {
+        continue;
+      }
+
+      candidates.push(i);
+      if (candidates.length >= need) {
+        break;
+      }
+    }
+    if (candidates.length === 0) {
+      return;
+    }
+
+    candidates.forEach((index) => {
+      this._uploadChunkRetryable(index).then(
+        () => {
+          this._uploadChunks();
+        },
+        (err) => {
+          this._onError(err);
+        }
+      );
+    });
+  }
+
+  _shouldRetry(err) {
+    if (err.source && typeof err.source.status === "number") {
+      const { status } = err.source;
+      // 0 probably indicates network failure
+      return (
+        status === 0 ||
+        status === 409 ||
+        status === 423 ||
+        (status >= 500 && status < 600)
+      );
+    }
+    return false;
+  }
+
+  _retryable({ before, attempt, after }) {
+    const { retryDelays } = this.options;
+    const { signal } = this.abortController;
+
+    if (before) {
+      before();
+    }
+
+    const doAttempt = (retryAttempt) =>
+      attempt().catch((err) => {
+        if (this._aborted()) {
+          throw createAbortError();
+        }
+
+        if (this._shouldRetry(err) && retryAttempt < retryDelays.length) {
+          return delay(retryDelays[retryAttempt], { signal }).then(() =>
+            doAttempt(retryAttempt + 1)
+          );
+        }
+        throw err;
+      });
+
+    return doAttempt(0).then(
+      (result) => {
+        if (after) {
+          after();
+        }
+        return result;
+      },
+      (err) => {
+        if (after) {
+          after();
+        }
+        throw err;
+      }
+    );
+  }
+
+  _uploadChunkRetryable(index) {
+    return this._retryable({
+      before: () => {
+        this.chunksInProgress += 1;
+      },
+      attempt: () => this._uploadChunk(index),
+      after: () => {
+        this.chunksInProgress -= 1;
+      },
+    });
+  }
+
+  _uploadChunk(index) {
+    this.chunkState[index].busy = true;
+
+    if (this._aborted()) {
+      this.chunkState[index].busy = false;
+      throw createAbortError();
+    }
+
+    return this._uploadChunkBytes(
+      index,
+      this.options.url,
+      this.options.headers
+    );
+  }
+
+  _onChunkProgress(index, sent) {
+    this.chunkState[index].uploaded = parseInt(sent, 10);
+
+    const totalUploaded = this.chunkState.reduce(
+      (total, chunk) => total + chunk.uploaded,
+      0
+    );
+    this.options.onProgress(totalUploaded, this.file.data.size);
+  }
+
+  _onChunkComplete(index) {
+    this.chunkState[index].done = true;
+    this.options.onChunkComplete(index);
+  }
+
+  _uploadChunkBytes(index, url, headers) {
+    const body = this.chunks[index];
+    const { signal } = this.abortController;
+
+    let defer;
+    const promise = new Promise((resolve, reject) => {
+      defer = { resolve, reject };
```

Comment:
Passing the scope upwards this way is a little unusual. Why not return `new Promise` from the function and do all your logic inside it?